apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-plugin/src/main/java/co/elastic/apm/agent/rabbitmq/ConnectionInstrumentation.java

```diff
@@ -0,0 +1,88 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.rabbitmq;
+
+import co.elastic.apm.agent.sdk.DynamicTransformer;
+import co.elastic.apm.agent.sdk.ElasticApmInstrumentation;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Connection;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.NamedElement;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.nameEndsWith;
+import static net.bytebuddy.matcher.ElementMatchers.nameStartsWith;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+/**
+ * Instruments
+ * <ul>
+ *     <li>{@link Connection#createChannel}</li>
+ * </ul>
+ */
+public class ConnectionInstrumentation extends BaseInstrumentation {
+
+    private static final Collection<Class<? extends ElasticApmInstrumentation>> CHANNEL_INSTRUMENTATIONS =
+        Arrays.<Class<? extends ElasticApmInstrumentation>>asList(
+            ChannelInstrumentation.BasicPublish.class,
+            ChannelInstrumentation.BasicConsume.class
+        );
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        // this fast heuristic works for all implementations within driver
+        return nameEndsWith("Connection")
+            .and(nameStartsWith("com.rabbitmq.client.impl"));
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        // fine to use super type matching thanks to restricting with pre-filter
+        return hasSuperType(named("com.rabbitmq.client.Connection"));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("createChannel");
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
+    public static void onExit(@Nullable @Advice.Thrown Throwable thrown,
+                              @Advice.Return @Nullable Channel channel) {
+
+        if (thrown != null || channel == null) {
+            return;
+        }
+        DynamicTransformer.Accessor.get().ensureInstrumented(channel.getClass(), CHANNEL_INSTRUMENTATIONS);
```

Comment:
I am not sure I understand why this is required instead of simple type matching for `Channel` implementations. 
However, since we already have this instrumentation, let's use it to store channel destination metadata (address and port) and use it to fill the send span destination details in the `basicPublish` instrumentation. Looks like very little is left in order to collect this info.

Reply:
Or what you did - use `com.rabbitmq.client.Channel#getConnection`, much simpler 😄